src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -263,7 +264,46 @@ private <T extends SyntaxElement> T parse(Iterator<? extends SyntaxElementInfo<?
 								x = x2;
 							}
 							T t = info.c.newInstance();
-							if (t.init(res.exprs, i, getParser().getHasDelayBefore(), res)) {
+							ParserInstance parser = getParser();
+							boolean error = false;
+							Class<? extends Structure>[] usableStructures = t.getUsableStructures();
+							if (parser.getCurrentStructure() != null) {
+								if (usableStructures.length > 0) {
+									error = true;
+									for (Class<? extends Structure> clazz : usableStructures) {
+										if (parser.isCurrentStructure(clazz)) {
+											error = false;
+											break;
+										}
+									}
+									if (error) {
+										Skript.error("'" + res.expr + "' cannot be used in " + Utils.a(parser.getCurrentStructure().toString(null, false)) + " structure.");
+									}
+								}
+							} else if (usableStructures.length > 0) {
+								error = true;
+								Skript.error("'" + res.expr + "' cannot be used outside of the allowed structures.");
+							}
```

Comment:
Very specific but what do you think about nesting this in 1 primary check like (basically same style as sections check below)
```java
if (usableStructures.length > 0) {
	error = true;
	if (parser.getCurrentStructure() == null) {
		// print error as there are restrictions but no current structure
	} else {
		// checks
	}
}
```

Reply:
Seems good to me c: